lazy-pages/src/sys/unix.rs

```diff
@@ -109,7 +109,102 @@ where
     }
 }
 
-pub unsafe fn setup_signal_handler<H>() -> io::Result<()>
+use errno::Errno;
+
+#[derive(Debug, Clone, Copy, derive_more::Display)]
+enum ThreadInitError {
+    #[display(fmt = "Cannot get information about old signal stack: {}", _0)]
+    OldStack(Errno),
+    #[display(fmt = "Cannot mmap space for signal stack: {}", _0)]
+    Mmap(Errno),
+    #[display(fmt = "Cannot set new signal stack: {}", _0)]
+    SigAltStack(Errno),
+}
+
+fn init_for_thread_internal() -> Result<(), ThreadInitError> {
+    use core::{mem, ptr};
+
+    // Should be enough for lazy-pages signal handler.
+    // Equal to libc::SIGSTKSZ on macos M1.
+    const SIGNAL_STACK_SIZE: usize = 0x20000;
+
+    enum StackInfo {
+        UseOldStack,
+        NewStack(*mut libc::c_void),
+    }
+
+    impl Drop for StackInfo {
+        fn drop(&mut self) {
+            if let StackInfo::NewStack(mmap_ptr) = self {
+                unsafe {
+                    // Deallocate the stack memory.
+                    if libc::munmap(*mmap_ptr, SIGNAL_STACK_SIZE) != 0 {
+                        log::error!(
+                            "Cannot deallocate signal stack memory during the thread shutdown: {}",
+                            errno::errno()
+                        );
+                    }
+                }
+            }
+        }
+    }
+
+    unsafe fn init_sigstack() -> Result<StackInfo, ThreadInitError> {
+        // Check whether old signal stack exist and suitable for lazy-pages signal handler.
+        let mut old_stack = mem::zeroed();
```

Comment:
```suggestion
        let mut old_stack = MaybeUninit::uninit();
```
Recommended to be used instead

Reply:
Done, check please